accelsim_tracing/sm/sm.go

```diff
@@ -0,0 +1,162 @@
+package sm
+
+import (
+	"log"
+
+	"github.com/sarchlab/akita/v3/sim"
+	"github.com/sarchlab/mgpusim/v3/accelsim_tracing/message"
+	"github.com/sarchlab/mgpusim/v3/accelsim_tracing/nvidia"
+	"github.com/sarchlab/mgpusim/v3/accelsim_tracing/subcore"
+)
+
+type SM struct {
+	*sim.TickingComponent
+
+	ID         string
+	warpsCount int64
+	instsCount int64
+
+	// meta
+	toGPU       sim.Port
+	toGPURemote sim.Port
+
+	toSubcores   sim.Port
+	Subcores     map[string]*subcore.Subcore
+	freeSubcores []*subcore.Subcore
+
+	undispatchedWarps    []*nvidia.Warp
+	unfinishedWarpsCount int64
+
+	finishedThreadblocksCount int64
+}
+
+func (s *SM) SetGPURemotePort(remote sim.Port) {
+	s.toGPURemote = remote
+}
+
+func (s *SM) Tick(now sim.VTimeInSec) bool {
+	madeProgress := false
+
+	madeProgress = s.reportFinishedKernels(now) || madeProgress
+	madeProgress = s.dispatchThreadblocksToSubcores(now) || madeProgress
+	madeProgress = s.processGPUInput(now) || madeProgress
+	madeProgress = s.processSubcoresInput(now) || madeProgress
+
+	// fmt.Println("SM tick, madeProgress:", madeProgress)
+
+	return madeProgress
+}
+
+func (s *SM) processGPUInput(now sim.VTimeInSec) bool {
+	msg := s.toGPU.Peek()
+	if msg == nil {
+		return false
+	}
+
+	switch msg := msg.(type) {
+	case *message.DeviceToSMMsg:
+		s.processSMMsg(msg, now)
+	default:
+		panic("Unhandled message type")
+	}
+
+	return true
+}
+
+func (s *SM) processSubcoresInput(now sim.VTimeInSec) bool {
+	msg := s.toSubcores.Peek()
+	if msg == nil {
+		return false
+	}
+
+	switch msg := msg.(type) {
+	case *message.SubcoreToSMMsg:
+		s.processSubcoreSubcoresg(msg, now)
+		return true
+	default:
+		panic("Unhandled message type")
+	}
+}
+
+func (s *SM) processSMMsg(msg *message.DeviceToSMMsg, now sim.VTimeInSec) {
+	for i := range msg.Threadblock.Warps {
+		s.undispatchedWarps = append(s.undispatchedWarps, &msg.Threadblock.Warps[i])
+		s.unfinishedWarpsCount++
+		s.warpsCount++
+	}
+
+	s.toGPU.Retrieve(now)
+}
+
+func (s *SM) processSubcoreSubcoresg(msg *message.SubcoreToSMMsg, now sim.VTimeInSec) {
+	if msg.WarpFinished {
+		s.freeSubcores = append(s.freeSubcores, s.Subcores[msg.SubcoreID])
+		s.unfinishedWarpsCount--
+		if s.unfinishedWarpsCount == 0 {
+			s.finishedThreadblocksCount++
+		}
+	}
+
+	s.toSubcores.Retrieve(now)
+}
+
+func (s *SM) reportFinishedKernels(now sim.VTimeInSec) bool {
```

Comment:
I think this func should be named with `reportFinishedThreadblocks`. 
SMs are not aware of the concept of `kernel`. It's not a big problem though.